components/service/glean/src/test/java/mozilla/components/service/glean/storages/MemoryDistributionsStorageEngineTest.kt

```diff
@@ -0,0 +1,279 @@
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.test.core.app.ApplicationProvider
+import kotlinx.coroutines.runBlocking
+import mozilla.components.service.glean.collectAndCheckPingSchema
+import mozilla.components.service.glean.private.Lifetime
+import mozilla.components.service.glean.private.MemoryDistributionMetricType
+import mozilla.components.service.glean.error.ErrorRecording
+import mozilla.components.service.glean.histogram.FunctionalHistogram
+import mozilla.components.service.glean.private.MemoryUnit
+import mozilla.components.service.glean.private.PingType
+import mozilla.components.service.glean.testing.GleanTestRule
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertTrue
+import org.junit.Assert.assertEquals
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.ArgumentMatchers
+import org.mockito.Mockito
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class MemoryDistributionsStorageEngineTest {
+
+    @get:Rule
+    val gleanRule = GleanTestRule(ApplicationProvider.getApplicationContext())
+
+    @Test
+    fun `accumulate() properly updates the values in all stores`() {
+        val storeNames = listOf("store1", "store2")
+
+        val metric = MemoryDistributionMetricType(
+            disabled = false,
+            category = "telemetry",
+            lifetime = Lifetime.Ping,
+            name = "test_memory_distribution",
+            sendInPings = storeNames,
+            memoryUnit = MemoryUnit.Kilobyte
+        )
+
+        // Create a sample that will fall into the underflow bucket (bucket '0') so we can easily
+        // find it
+        val sample = 1L
+        MemoryDistributionsStorageEngine.accumulate(
+            metricData = metric,
+            sample = sample,
+            memoryUnit = MemoryUnit.Kilobyte
+        )
+
+        // Check that the data was correctly set in each store.
+        for (storeName in storeNames) {
+            val snapshot = MemoryDistributionsStorageEngine.getSnapshot(
+                storeName = storeName,
+                clearStore = true
+            )
+            assertEquals(1, snapshot!!.size)
+            assertEquals(1L, snapshot["telemetry.test_memory_distribution"]?.values!![1024])
+        }
+    }
+
+    @Test
+    fun `deserializer should correctly parse memory distributions`() {
+        val td = FunctionalHistogram()
+
+        val persistedSample = mapOf(
+            "store1#telemetry.invalid_string" to "invalid_string",
+            "store1#telemetry.invalid_bool" to false,
+            "store1#telemetry.null" to null,
+            "store1#telemetry.invalid_int" to -1,
+            "store1#telemetry.invalid_list" to listOf("1", "2", "3"),
+            "store1#telemetry.invalid_int_list" to "[1,2,3]",
+            "store1#telemetry.invalid_td_values" to "{\"values\":{\"0\": \"nope\"},\"sum\":0}",
```

Comment:
nit: The `td` in `invalid_td_values` here probably referred to timing distribution...